gateway/ln-gateway/src/db.rs

```diff
@@ -508,6 +523,82 @@ async fn migrate_to_v4(mut ctx: MigrationContext<'_>) -> Result<(), anyhow::Erro
     Ok(())
 }
 
+async fn migrate_to_v5(mut ctx: MigrationContext<'_>) -> Result<(), anyhow::Error> {
+    let mut dbtx = ctx.dbtx();
+    migrate_federation_config(&mut dbtx).await
+}
+
+async fn migrate_federation_config(
+    dbtx: &mut DatabaseTransaction<'_>,
+) -> Result<(), anyhow::Error> {
+    async fn migrate_client_entries(
+        dbtx: &mut DatabaseTransaction<'_>,
+        prefix: &[u8],
+    ) -> Result<(), anyhow::Error> {
+        let isolated_entries = dbtx
+            .raw_find_by_prefix(prefix)
+            .await?
+            .collect::<BTreeMap<_, _>>()
+            .await;
+        for (mut key, value) in isolated_entries {
+            dbtx.raw_remove_entry(&key).await?;
+            let mut new_key = vec![DbKeyPrefix::ClientDatabase as u8];
+            new_key.append(&mut key);
+            dbtx.raw_insert_bytes(&new_key, &value).await?;
+        }
+
+        Ok(())
+    }
+
+    // We need to migrate all isolated database entries to be behind the 0x10
+    // prefix. The problem is, if there is a `FederationId` that starts with
+    // 0x04, we cannot read the `FederationId` because the database will be confused
+    // between the isolated DB and the `FederationIdKey` record. To solve this,
+    // we first try and see if there are any entries that begin with
+    // 0x0404. This indicates a joined federation that will have a
+    // problem decoding the ID.
+    let problem_fed_configs = dbtx
+        .raw_find_by_prefix(&[0x04, 0x04])
+        .await?
+        .collect::<BTreeMap<_, _>>()
+        .await;
+    for (problem_key, problem_fed_config) in problem_fed_configs {
+        let federation_id = FederationId::consensus_decode_vec(
+            problem_key[1..33].to_vec(),
+            &ModuleDecoderRegistry::default(),
+        )?;
+        tracing::warn!(
+            ?federation_id,
+            "Found a FederationConfig entry that will cause issues decoding"
+        );
+        let federation_id_bytes = problem_key[1..3].to_vec();
+
+        // Remove the `FederationConfig` entry so that the migration of the isolated
+        // database doesn't accidentally migrate that too.
+        dbtx.raw_remove_entry(&problem_key).await?;
+
+        migrate_client_entries(dbtx, &federation_id_bytes).await?;
+
+        // Re-insert the `FederationConfig` entry so that it exists after the migration.
+        dbtx.raw_insert_bytes(&problem_key, &problem_fed_config)
+            .await?;
+    }
+
+    // Migrate the rest of the isolated databases that don't overlap with
+    // `FederationConfig`
+    let fed_ids = dbtx
+        .find_by_prefix(&FederationIdKeyPrefix)
```

Comment:
This prefix ... is empty. Doesn't it ... match everything in the database? I'm confused.

Reply:
Hmm...

```rust
impl_db_record!(
    key = FederationIdKey,
    value = FederationConfig,
    db_prefix = DbKeyPrefix::FederationConfig,
);
```

The naming here is not ideal. SHouldn't db_prefix and key match in the names?

`s/FederationIdKey/FederationConfigKey/`? The `FederationIdKey` _contains_ federation id, but it's actually a key to all `FederationConfig`s, right?

Also `s/FederationIdKeyPrefix/FederationConfigPrefixAll` or something?=

Reply:
OK, so this is actually the bulk of the migration:

"Read all the federation ids using `FederationConfig`s, and migrate keys that match to the correct version". Might update the comment as "Migrate the rest" gives the impression like this is some kind of "minor part".

But given that the federation config was

Reply:
The `impl_db_lookup` will automatically add the `0x04` prefix (defined in the `DbKeyPrefix` enum). So it will be all entries beginning with `0x04`

Reply:
Yeah we need to use the `FederationConfig` records that are indexed by the `FederationID` to know what the IDs are. Agree that the naming here is not ideal. Would like to leave that to a separate improvement though

Reply:
Renamed `FederationIdKey` to `FederationConfigKey`